// vscroll/dist/esm5/processes/misc/enums.js
export var CommonProcess;
(function (CommonProcess) {
  CommonProcess["init"] = "init";
  CommonProcess["scroll"] = "scroll";
  CommonProcess["start"] = "start";
  CommonProcess["preFetch"] = "preFetch";
  CommonProcess["fetch"] = "fetch";
  CommonProcess["postFetch"] = "postFetch";
  CommonProcess["render"] = "render";
  CommonProcess["preClip"] = "preClip";
  CommonProcess["clip"] = "clip";
  CommonProcess["adjust"] = "adjust";
  CommonProcess["end"] = "end";
})(CommonProcess || (CommonProcess = {}));
export var AdapterProcess;
(function (AdapterProcess) {
  AdapterProcess["reset"] = "adapter.reset";
  AdapterProcess["reload"] = "adapter.reload";
  AdapterProcess["append"] = "adapter.append";
  AdapterProcess["prepend"] = "adapter.prepend";
  AdapterProcess["check"] = "adapter.check";
  AdapterProcess["remove"] = "adapter.remove";
  AdapterProcess["replace"] = "adapter.replace";
  AdapterProcess["update"] = "adapter.update";
  AdapterProcess["clip"] = "adapter.clip";
  AdapterProcess["insert"] = "adapter.insert";
  AdapterProcess["fix"] = "adapter.fix";
})(AdapterProcess || (AdapterProcess = {}));
export var ProcessStatus;
(function (ProcessStatus) {
  ProcessStatus["start"] = "start";
  ProcessStatus["next"] = "next";
  ProcessStatus["done"] = "done";
  ProcessStatus["error"] = "error";
})(ProcessStatus || (ProcessStatus = {}));
